src/storage/src/hummock/cache.rs

```diff
@@ -0,0 +1,692 @@
+            let old = self.table.insert(hash, ptr);
+            if !old.is_null() {
+                let data = self.remove_cache_handle(old).unwrap();
+                last_reference_list.push(data);
+            }
+            self.usage += charge;
+            (*ptr).add_ref();
+            ptr
+        }
+    }
+
+    unsafe fn release(&mut self, e: *mut LruHandle<K, T>) -> Option<T> {
+        if e.is_null() {
+            return None;
+        }
+        let last_reference = (*e).unref();
+        if last_reference && (*e).is_in_cache() {
+            if self.usage > self.capacity {
+                self.table.remove((*e).hash, &(*e).key);
+                (*e).set_in_cache(false);
+            } else {
+                self.lru_insert(e);
+                return None;
+            }
+        }
+        if last_reference {
+            let mut node = Box::from_raw(e);
+            self.usage -= node.charge;
+            let data = node.value.take().unwrap();
+            self.recyle_handle_object(node);
+            Some(data)
+        } else {
+            None
+        }
+    }
+
+    unsafe fn lookup(&mut self, hash: u64, key: &K) -> *mut LruHandle<K, T> {
+        let e = self.table.lookup(hash, key);
+        if !e.is_null() {
+            if !(*e).has_refs() {
+                self.lru_remove(e);
+            }
+            (*e).add_ref();
+        }
+        e
+    }
+
+    unsafe fn erase(&mut self, hash: u64, key: &K) -> Option<T> {
+        let h = self.table.remove(hash, key);
+        if !h.is_null() {
+            self.remove_cache_handle(h)
+        } else {
+            None
+        }
+    }
+
+    unsafe fn remove_cache_handle(&mut self, h: *mut LruHandle<K, T>) -> Option<T> {
+        (*h).set_in_cache(false);
+        if !(*h).has_refs() {
+            self.lru_remove(h);
+            let mut node = Box::from_raw(h);
+            node.set_in_cache(false);
+            self.usage -= node.charge;
+            let data = node.value.take();
+            self.recyle_handle_object(node);
+            return data;
+        }
+        None
+    }
+}
+
+pub struct LruCache<K: PartialEq + Default, T> {
+    num_shard_bits: usize,
+    shards: Vec<Mutex<LruCacheShard<K, T>>>,
+}
+
+impl<K: PartialEq + Default, T> LruCache<K, T> {
+    pub fn new(num_shard_bits: usize, capacity: usize, object_cache: usize) -> Self {
+        let num_shards = 1 << num_shard_bits;
+        let mut shards = Vec::with_capacity(num_shards);
+        let per_shard = capacity / num_shards;
+        let per_shard_object = object_cache / num_shards;
+        for _ in 0..num_shards {
+            shards.push(Mutex::new(LruCacheShard::new(
+                per_shard,
+                per_shard_object,
+                false,
+            )));
+        }
+        Self {
+            num_shard_bits,
+            shards,
+        }
+    }
+
+    pub fn lookup(self: &Arc<Self>, hash: u64, key: &K) -> Option<CachableEntry<K, T>> {
```

Comment:
Why lookup needs to provide a hash? Users are expected to calculate hash by themselves?

Reply:
I will use a BlockCache to wrapper  Cache<(u64,u64), Box<Block>> and provide a hash algorithm there.